codemp/rd-rend2/glsl/ssao.glsl

```diff
@@ -8,95 +8,92 @@ void main()
-	float sinr = sin(r);
-	float cosr = cos(r);
-	return mat2(cosr, sinr, -sinr, cosr);
+float compareDepths( in float depth1, in float depth2 ) {
+	float diff = sqrt( clamp(1.0-(depth1-depth2) / ( u_SSAOSettings2.y /* aorange */ / (u_ViewInfo.y - u_ViewInfo.x)),0.0,1.0) );
+	float ao = min(u_SSAOSettings.x /* aocap */,max(0.0,depth1-depth2-u_SSAOSettings2.z /* depthTolerance */) * u_SSAOSettings.z /* aoMultiplier */) * diff;
+	return ao;
 }
 
-float getLinearDepth(sampler2D depthMap, const vec2 tex, const float zFarDivZNear)
+vec2 rand(vec2 coord) //generating random noise
 {
-		float sampleZDivW = texture(depthMap, tex).r;
-		return 1.0 / mix(zFarDivZNear, 1.0, sampleZDivW);
+	float noiseX = (fract(sin(dot(coord ,vec2(12.9898,78.233))) * 43758.5453));
+	float noiseY = (fract(sin(dot(coord ,vec2(12.9898,78.233)*2.0)) * 43758.5453));
+	return vec2(noiseX,noiseY)*0.004;
 }
 
-float ambientOcclusion(sampler2D depthMap, const vec2 tex, const float zFarDivZNear, const float zFar)
+void main(void)
 {
-	float result = 0;
-
-	float sampleZ = zFar * getLinearDepth(depthMap, tex, zFarDivZNear);
-
-	vec2 expectedSlope = vec2(dFdx(sampleZ), dFdy(sampleZ)) / vec2(dFdx(tex.x), dFdy(tex.y));
-	
-	if (length(expectedSlope) > 5000.0)
-		return 1.0;
-	
-	vec2 offsetScale = vec2(3.0 / sampleZ);
-	
-	mat2 rmat = randomRotation(tex);
-		
+	float depth = readDepth( var_ScreenTex );
+	vec2 noise = rand( var_ScreenTex );
+
+	float d;
+ 
+	float pw = 1.0 / u_ScreenInfo.x;
+	float ph = 1.0 / u_ScreenInfo.y;
+	if (u_SSAOSettings2.x > 0) // apply random noise?
+	{
+		pw /= clamp(depth,0.05,1.0)+(noise.x*(1.0-noise.x));
+		ph /= clamp(depth,0.05,1.0)+(noise.y*(1.0-noise.y));
+	}
+ 
+	float aoCap = u_SSAOSettings.x;
+	float ao = 0.0;
+	float aoScale = 1.0;
+
 	int i;
-	for (i = 0; i < 3; i++)
+	for (i = 0; i < 4; i++)
 	{
-		vec2 offset = rmat * poissonDisc[i] * offsetScale;
-		float sampleZ2 = zFar * getLinearDepth(depthMap, tex + offset, zFarDivZNear);
-
-		if (abs(sampleZ - sampleZ2) > 20.0)
-			result += 1.0;
-		else
-		{
-			float expectedZ = sampleZ + dot(expectedSlope, offset);
-			result += step(expectedZ - 1.0, sampleZ2);
-		}
+		d = readDepth( vec2(var_ScreenTex.x+pw,var_ScreenTex.y+ph));
+		ao += compareDepths(depth,d) / aoScale;
+
+		d = readDepth( vec2(var_ScreenTex.x-pw,var_ScreenTex.y+ph));
+		ao += compareDepths(depth,d) / aoScale;
+
+		d = readDepth( vec2(var_ScreenTex.x+pw,var_ScreenTex.y-ph));
+		ao += compareDepths(depth,d) / aoScale;
+
+		d = readDepth( vec2(var_ScreenTex.x-pw,var_ScreenTex.y-ph));
+		ao += compareDepths(depth,d) / aoScale;
+
+		pw *= 2.0;
+		ph *= 2.0;
+		aoScale *= 1.2;
 	}
-	
-	result *= 0.33333;
-	
-	return result;
-}
 
-void main()
-{
-	float result = ambientOcclusion(u_ScreenDepthMap, var_ScreenTex, u_ViewInfo.x, u_ViewInfo.y);
-			
-	out_Color = vec4(vec3(result), 1.0);
+	ao /= 16.0;
+	ao *= u_SSAOSettings.y;
+ 
+	float done = (1.1 - ao);
+	if (u_SSAOSettings.w > 1)
+	{
+		float orig = texture2D(u_ScreenImageMap,var_ScreenTex).x;
```

Comment:
`texture2D` should be `texture`.

But also, which texture is this reading? Why does a SSAO shader need anything other than the screen's depth texture?